### DISCUSSION TITLE
Save State

### DISCUSSION THREAD

**Discussion by adasi11** [rsachetto/MonoAlg3D_C#45] [Ideas]
Hey, 

Here Albert,

I was wondering whether it would be possible to "create" a state that can be restored for future simulations, but without running the previous simulation and saving state. That is, providing that I have the voltage values for every node and all parameters for the simulation, can I create a "mock" state that I can restore?   

Thanks

**Comment by rsachetto** [comment]
Hi Albert, it is possible to do that. You will basically need all state variables (including Voltage) for the ODEs of each node. There is a function in matrix_assembly.c called set_initial_conditions_from_odes. This function sets the tissue voltage according to the ODEs voltage.  So, you will have to use this function as the init_function in the [assembly_matrix] section.

Now, the hardest part is to restore the ODEs values. In order to do that, you will have to create an extra_data function that will assemble 
the SV vector. This vector is assembled using the following scheme, considering a model with Voltage, 3 other state variables, and 3 volumes:

 [V1, V2, V3, S1_1, S1_2, S1_3, S2_1, S2_2, S2_3, S3_1, S3_2, S3_3]

where V1, V2, V3 are the voltage values for the 3 cells. Si_j is the state variable i for cell j.

After returning the vector in the extra_data function, you can write the set_model_initial_conditions_gpu as:

```c
extern "C" SET_ODE_INITIAL_CONDITIONS_GPU(set_model_initial_conditions_gpu) {

    uint32_t num_volumes = solver->original_num_cells;

    // execution configuration
    const int GRID  = (num_volumes + BLOCK_SIZE - 1)/BLOCK_SIZE;

    size_t size = num_volumes*sizeof(real);

    check_cuda_error(cudaMallocPitch((void **) &(solver->sv), &pitch_h, size, (size_t )NEQ));
    check_cuda_error(cudaMemcpyToSymbol(pitch, &pitch_h, sizeof(size_t)));

    real *initial_conditions = NULL;
    real *initial_conditions_device = NULL;

    if(solver->ode_extra_data) {
        initial_conditions = (real *)solver->ode_extra_data;
        check_cuda_error(cudaMemcpy2D (solver->sv, pitch_h, initial_conditions, size, size, (size_t) NEQ, cudaMemcpyHostToDevice));

    }
    else {
        kernel_set_model_inital_conditions <<<GRID, BLOCK_SIZE>>>(solver->sv, num_volumes);
    }

    check_cuda_error( cudaPeekAtLastError() );
    cudaDeviceSynchronize();

    check_cuda_error(cudaFree(initial_conditions_device));

    return pitch_h;

}
```

I hope this helps. Let me know if you have any further questions.

My best.

**Reply by adasi11** [reply]
Okey, 
This seems doable. 
Now my question is: Will I still be able to use other functions in [assembly_matrix], such as anisotropic_sigma_assembly_matrix? Or other parameters in [extra_data] such as tissue_type? I guess the latter depends on how I create the [extra_data] function to include everything. I am more concerned about [assembly_matrix].

**Reply by rsachetto** [reply]
> Okey, This seems doable. Now my question is: Will I still be able to use other functions in [assembly_matrix], such as anisotropic_sigma_assembly_matrix? 

Yes. Only the init_function will change.

Or other parameters in [extra_data] such as tissue_type?
You can pass any parameters that you want to extra_data. 

 I guess the latter depends on how I create the [extra_data] function to include everything. I am more concerned about [assembly_matrix].
You don't need to change assembly_matrix function.

Let me know i I can help.

My best,
Rafael Sachetto

**Reply by adasi11** [reply]
Perfect! 
For now I will give it a try. But I will get in touch if needed. Thanks!

It is my uderstanding that there's a function for saving the state variables of one element of the meshover time. 
But, is there a function to save the state variables of all elements in the mesh at a given time step? 

Best,
Albert

**Reply by rsachetto** [reply]
Take a look in the save_state.c file, starting at line 88. In this function, I will dump all odes in a file called ode_solver_checkpoint.dat. It saves more than just the state variables, but I think it is a good example to start.

**Reply by rsachetto** [reply]
After this discussion, I ended up adding the option to save the state variables when using the ensight file format. In order to enable this, just add the option save_ode_state_variables=true in the save_result section, when using the save_as_ensight function.

Take a look in the save_as_ensigth function in src/save_mesh/save_mesh.c and the function save_en6_file_state_vars in  src/ensight_utils/ensight_grid.h .

Hope this helps,

Rafael

**Reply by adasi11** [reply]
That's amazing!

Many thanks, Rafael

**Comment by adasi11** [comment]
Hey Rafael,

Do you think it would be better to pass the vector of state variables as a text file? Or include the state variables as different columns in the ".alg" file and generate a function to enssemble the vector later? 

Thanks,
ALbert

**Comment by rsachetto** [comment]
Hi Abert,
In my opinion, it would be easy to save the vector of state variables as a "matrix", where each state var is a line in the file.
So you can easily create the sv by reading each line (where the number of columns is the number of active cells of the mesh).

Hope it helps, 

My best,
Rafael Sachetto

**Reply by adasi11** [reply]
Hey Rafael,

I also want to include tissue type and 13 additional parameters in [extra_data], so maybe I should read the "matrix" of state variables in "set_model_initial_conditions_gpu", without using the [extra_data] function?

**Reply by rsachetto** [reply]
You can do that as well. Usually, I crate a struct in the extra data code to handle all the data that I need to pass. But you can also read the file in the set_model_initial_conditions_gpu part.

my best,
Rafael

**Reply by adasi11** [reply]
May I have an example of the extra_data struct? That would be very helpful!
Thanks!

Albert

**Reply by rsachetto** [reply]
There is a struct called extra_data_for_fibrosis in src/extra_data_library/helper_functions.h that is used in set_extra_data_for_fibrosis_plain and is set in set_common_schemia_data.

You can create something similat and add everything you need in it and cast it back in the model, like I do in src/models_library/ten_tusscher/ten_tusscher_3_RS_GPU.cu:

struct extra_data_for_fibrosis* extra_data_from_cpu = (struct extra_data_for_fibrosis*)ode_solver->ode_extra_data;

Let me know if it helps.

My best,
Rafael Sachetto

**Comment by adasi11** [comment]
Could you tell me your thoughts about the following struct?

I am reading the state variables from a column array in the form: 
[ V1_1
  V1_2
  ...
  V1_n
  V2_1
  ...
  V2_n
  ...
  Vm_n];

n = num_active_cells
m = NEQ

**Reply by adasi11** [reply]
struct extra_data_for_fibrillation {
    real GKur;
    real GKr;
    real Gto;
    real GK1;
    real GKs;
    real GCaL;
    real GNaK;
    real GNCX;
    real GNa;
    real Gup;
    real Grel;
    real Gleak;
    real Drug;
    real *tissue_type;
    real *state_var;
};
struct extra_data_for_fibrillation * set_data_sv_restoration(struct config *config, uint32_t num_cells);



struct extra_data_for_fibrillation* set_data_sv_restoration(struct config *config, uint32_t num_cells) {

    struct extra_data_for_fibrillation *extra_data = MALLOC_ONE_TYPE(struct extra_data_for_fibrillation);
    size_t num_sv_entries = 25; // NEQ

    real GKur = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GKur, config, "GKur");
    real GKr = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GKr, config, "GKr");
    real Gto = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, Gto, config, "Gto");
    real GK1 = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GK1, config, "GK1");
    real GKs = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GKs, config, "GKs");
    real GCaL = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GCaL, config, "GCaL");
    real GNaK = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GNaK, config, "GNaK");
    real GNCX = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GNCX, config, "GNCX");
    real GNa = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, GNa, config, "GNa");
    real Gup = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, Gup, config, "Gup");
    real Grel = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, Grel, config, "Grel");
    real Gleak = 1.0f;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, Gleak, config, "Gleak");
    real Drug = 0;
    GET_PARAMETER_NUMERIC_VALUE_OR_USE_DEFAULT(real, Drug, config, "Drug");

    extra_data->GKur = GKur;
    extra_data->GKr = GKr;
    extra_data->Gto = Gto;
    extra_data->GK1 = GK1;
    extra_data->GKs = GKs;
    extra_data->GCaL = GCaL;
    extra_data->GNaK = GNaK;
    extra_data->GNCX = GNCX;
    extra_data->GNa = GNa;
    extra_data->Gup = Gup;
    extra_data->Grel = Grel;
    extra_data->Gleak = Gleak;
    extra_data->Drug = Drug;
    extra_data->tissue_type = MALLOC_ARRAY_OF_TYPE(real, num_cells);
    extra_data->state_var = MALLOC_ARRAY_OF_TYPE(real, num_cells * num_sv_entries);

    return extra_data;
}



SET_EXTRA_DATA(set_extra_data_fibrillation) {

    uint32_t num_active_cells = the_grid->num_active_cells;

    struct extra_data_for_fibrillation *extra_data = NULL;
    extra_data = set_data_sv_restoration(config, num_active_cells);

    struct cell_node ** ac = the_grid->active_cells;

    OMP(parallel for)
    for(uint32_t i = 0; i < num_active_cells; i++) {
        extra_data->tissue_type[i] = SET_TISSUE_TYPE(ac[i]);;
    }

    char *sv_file = NULL;
    GET_PARAMETER_STRING_VALUE_OR_REPORT_ERROR(sv_file, config, "sv_file");

    FILE *statevar_file = fopen(StateVar_file, "r");
    if(!statevar_file) {
        fprintf(stderr, "Error opening state variable file %s! Exiting!\n", statevar_file);
        exit(EXIT_FAILURE);
    }

    uint32_t num_sv_cells = num_active_cells * 25;
    for(int i = 0; i < num_sv_cells; i++) {
        fscanf(sv_file, "%lf\n", &extra_data->state_var[i]);
    }
    return (void*)extra_data;
}

**Reply by adasi11** [reply]
With the following modification in the function that you sent me:


real *initial_conditions = NULL;
real *initial_conditions_device = NULL;

struct extra_data_for_fibrillation* extra_data_from_cpu = (struct extra_data_for_fibrillation*)ode_solver->ode_extra_data;

if(solver->ode_extra_data) {
        initial_conditions = extra_data_from_cpu->state_var;
        check_cuda_error(cudaMemcpy2D (solver->sv, pitch_h, initial_conditions, size, size, (size_t) NEQ, cudaMemcpyHostToDevice));
}

**Reply by rsachetto** [reply]
Hi Albert, it seems correct. You just forgot to set the size before returning:

SET_EXTRA_DATA_SIZE(sizeof(struct extra_data_for_fibrillation));

edit: I think you will not need the size here. But is safer to set it.

**Reply by rsachetto** [reply]
One suggestion that I have is to add the num_sv_entries as a config parameter, so you can easily use this with different cell models.

**Reply by adasi11** [reply]
Thanks!

**Reply by adasi11** [reply]
One more thing (hopefully the last one).
Should I pass as input the actual "dt", "time_new" and "previous_dt"? Or can I set these variables to "min_dt", "0" and "0", respectively?

**Reply by rsachetto** [reply]
If you don't need to keep track of them you can keep them like in the other models. Your simulation will always start in time 0 but will use the initial condition that you read from the file.
